File: app/src/main/java/jp/kentan/student_portal_plus/ui/span/CustomTabsURLSpan.java
package jp.kentan.student_portal_plus.ui.span;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.support.customtabs.CustomTabsIntent;
import android.support.v4.content.ContextCompat;
import android.text.style.URLSpan;
import android.view.View;

import org.chromium.customtabsclient.shared.CustomTabsHelper;

import jp.kentan.student_portal_plus.R;

@SuppressLint("ParcelCreator")
public class CustomTabsURLSpan extends URLSpan {

    private final SharedPreferences mPreference;
    private final String GOOGLE_DOCS_URL;

    public CustomTabsURLSpan(Context context, String url) {
        super(url);
        mPreference = context.getSharedPreferences("common", Context.MODE_PRIVATE);
        GOOGLE_DOCS_URL = context.getString(R.string.url_gdocs);
    }

//    public CustomTabsURLSpan(Context context, Parcel src) {
//        super(src);
//        this.mPreference = context.getSharedPreferences("common", Context.MODE_PRIVATE);
//    }

    @Override
    public void onClick(View view) {
        String url = this.getURL();

        if (mPreference.getBoolean("pdf_open_with_gdocs", true) && url.endsWith(".pdf")) {
            url = GOOGLE_DOCS_URL + url;
        }

        String packageName = CustomTabsHelper.getPackageNameToUse(view.getContext());

        CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();

        builder.setShowTitle(true);
        builder.addDefaultShareMenuItem();
        builder.setToolbarColor(ContextCompat.getColor(view.getContext(), R.color.colorPrimary));

        CustomTabsIntent customTabsIntent = builder.build();
        customTabsIntent.intent.setPackage(packageName);
        customTabsIntent.launchUrl(view.getContext(), Uri.parse(url));
    }
}

File: app/src/main/java/jp/kentan/studentportalplus/ui/span/CustomTabsUrlSpan.kt
package jp.kentan.studentportalplus.ui.span

import android.content.Context
import android.content.Intent
import android.support.customtabs.CustomTabsIntent
import android.support.v4.content.ContextCompat
import android.text.style.URLSpan
import android.view.View
import androidx.core.net.toUri
import jp.kentan.studentportalplus.R
import jp.kentan.studentportalplus.util.CustomTabsHelper
import jp.kentan.studentportalplus.util.enabledPdfOpenWithGdocs
import org.jetbrains.anko.defaultSharedPreferences
import org.jetbrains.anko.longToast

class CustomTabsUrlSpan(private val context: Context, url: String) : URLSpan(url) {

    override fun onClick(widget: View?) {
        val isPdf = url.endsWith(".pdf", true)
        val isRequireLogin = url.startsWith("https://portal.student.kit.ac.jp", true)

        var urlStr = url
        if (isPdf && context.defaultSharedPreferences.enabledPdfOpenWithGdocs()) {
            if (isRequireLogin) {
                context.longToast(R.string.error_gdocs_require_login)
            } else {
                urlStr = context.getString(R.string.url_gdocs, url)
            }
        }

        val customTabs = CustomTabsIntent.Builder()
                .setShowTitle(true)
                .addDefaultShareMenuItem()
                .setToolbarColor(ContextCompat.getColor(context, R.color.colorPrimary))
                .build()

        customTabs.intent.`package` = CustomTabsHelper.getPackageNameToUse(context)
        customTabs.intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        customTabs.launchUrl(context, urlStr.toUri())
    }
}